javascore/nativecoinIRC2/src/main/java/foundation/icon/btp/nativecoinIRC2/NativeCoinService.java

```diff
@@ -0,0 +1,579 @@
+    private final DictDB<BigInteger, TransferTransaction> transactions = Context.newDictDB("transactions", TransferTransaction.class);
+
+    public NativeCoinService(Address _bmc, Address _irc2, String _name, String _tokenName) {
+        bmc = _bmc;
+        BMCScoreInterface bmcInterface = new BMCScoreInterface(bmc);
+        //BTPAddress btpAddress = BTPAddress.valueOf(bmcInterface.getBtpAddress());
+        String btpAddr = (String) Context.call(bmcInterface._getAddress(), "getBtpAddress");
+        BTPAddress btpAddress = BTPAddress.valueOf(btpAddr);
+        net = btpAddress.net();
+        irc2 = _irc2;
+        name = _name;
+        tokenName = _tokenName;
+        coinNames.add(_tokenName);
+    }
+
+    public NCSProperties getProperties() {
+        return properties.getOrDefault(NCSProperties.DEFAULT);
+    }
+
+    public void setProperties(NCSProperties properties) {
+        this.properties.set(properties);
+    }
+
+    private boolean isRegistered(String name) {
+        if (tokenName.equals(name)) {
+            return true;
+        }
+        return false;
+    }
+
+    static void require(boolean condition, String message) {
+        if (!condition) {
+            throw NCSException.unknown(message);
+        }
+    }
+
+    @External(readonly = true)
+    public String[] coinNames() {
+        String[] names = new String[2];
+        names[0] = name;
+        names[1] = tokenName;
+        return names;
+    }
+
+    @External(readonly = true)
+    public Balance balanceOf(Address _owner, String _coinName) {
+        if (_owner.equals(Context.getAddress())) {
+            Balance balance = new Balance(BigInteger.ZERO, BigInteger.ZERO, BigInteger.ZERO);
+            balance.setRefundable(feeBalances.getOrDefault(_coinName, BigInteger.ZERO));
+            return balance;
+        } else {
+            return getBalance(_coinName, _owner);
+        }
+    }
+
+    @External
+    public void reclaim(String _coinName, BigInteger _value) {
+        require(_value.compareTo(BigInteger.ZERO) > 0, "_value must be positive");
+
+        Address owner = Context.getCaller();
+        Balance balance = getBalance(_coinName, owner);
+        require(balance.getRefundable().compareTo(_value) > -1, "invalid value");
+        balance.setRefundable(balance.getRefundable().subtract(_value));
+        setBalance(_coinName, owner, balance);
+
+        if (name.equals(_coinName)) {
+            Context.transfer(owner, _value);
+        } else {
+            transfer(irc2, owner, _value, "Reclaim: transferring to Caller".getBytes());
+        }
+    }
+
+    @External
+    public void tokenFallback(Address from, BigInteger value, @Optional byte[] data) {
+        require(value.compareTo(BigInteger.ZERO) > 0, "Invalid Amount");
+        require(Context.getCaller().equals(irc2), "Invalid Caller: Token Contract wrong");
+        require(tokenName != null && !tokenName.equals(""), "Token not registered");
```

Comment:
what is the purpose of this line?

Reply:
ERC20 in javascore handles token deposit to contract using tokenFallback mechanism

Reply:
are there any cases that tokenName can be null?

Reply:
Good catch, this line was carried over from multi token support from Token-transfer BSH. removed it now. Thanks